File: packages/skill-eval/src/commands/validate.ts
import fs from "node:fs";
import pc from "picocolors";
import { loadConfig, resolveSkillPaths } from "../config.js";
import { EvalsFileSchema } from "../schemas/evals.js";
import { extractErrorMessage } from "../utils/error.js";

interface ValidateOpts {
	skill: string;
	evals?: string;
	config?: string;
}

export async function runValidate(opts: ValidateOpts) {
	let hasError = false;

	function pass(msg: string) {
		console.log(`  ${pc.green("✓")} ${msg}`);
	}
	function fail(msg: string) {
		console.log(`  ${pc.red("✗")} ${msg}`);
		hasError = true;
	}

	console.log(pc.bold("\nSkill Eval — Validate\n"));

	// Skill directory
	const paths = resolveSkillPaths(opts.skill, opts.evals);

	if (fs.existsSync(paths.skillDir) && fs.statSync(paths.skillDir).isDirectory()) {
		pass(`Skill directory: ${paths.skillDir}`);
	} else {
		fail(`Skill directory not found: ${paths.skillDir}`);
	}

	// SKILL.md
	if (fs.existsSync(paths.skillFile)) {
		const size = fs.statSync(paths.skillFile).size;
		pass(`SKILL.md found (${(size / 1024).toFixed(1)} KB)`);
	} else {
		fail(`SKILL.md not found at ${paths.skillFile}`);
	}

	// evals.json
	if (fs.existsSync(paths.evalsFile)) {
		try {
			const raw = JSON.parse(fs.readFileSync(paths.evalsFile, "utf-8"));
			const parsed = EvalsFileSchema.parse(raw);
			const totalAssertions = parsed.evals.reduce((sum, e) => sum + e.assertions.length, 0);
			pass(`evals.json valid (${parsed.evals.length} evals, ${totalAssertions} assertions)`);
		} catch (err) {
			fail(`evals.json invalid: ${extractErrorMessage(err)}`);
		}
	} else {
		fail(`evals.json not found at ${paths.evalsFile}`);
	}

	// Config / providers
	try {
		const config = loadConfig({ configPath: opts.config });
		for (const p of config.providers) {
			pass(`${p.model}: API key valid`);
		}
	} catch (err) {
		fail(`Config: ${extractErrorMessage(err)}`);
	}

	console.log("");

	if (hasError) {
		console.log(pc.red("Validation found issues. Fix them before running evals.\n"));
		process.exit(1);
	} else {
		console.log(pc.green("All checks passed. Ready to run evals.\n"));
	}
}

File: packages/skill-eval-python/src/skill_eval/commands/validate.py
from __future__ import annotations

import json
from pathlib import Path

from rich.console import Console

from skill_eval.config import load_config, resolve_skill_paths
from skill_eval.config.schema import CliOverrides
from skill_eval.schemas.evals import EvalsFile
from skill_eval.utils.error import extract_error_message


async def run_validate(
    skill: str,
    evals_path: str | None = None,
    config_path: str | None = None,
) -> None:
    console = Console()
    has_error = False

    def pass_msg(msg: str) -> None:
        console.print(f"  [green]✓[/green] {msg}")

    def fail_msg(msg: str) -> None:
        nonlocal has_error
        console.print(f"  [red]✗[/red] {msg}")
        has_error = True

    console.print("\n[bold]Skill Eval — Validate[/bold]\n")

    paths = resolve_skill_paths(skill, evals_path)

    if Path(paths["skill_dir"]).exists() and Path(paths["skill_dir"]).is_dir():
        pass_msg(f"Skill directory: {paths['skill_dir']}")
    else:
        fail_msg(f"Skill directory not found: {paths['skill_dir']}")

    skill_file = Path(paths["skill_file"])
    if skill_file.exists():
        size_kb = skill_file.stat().st_size / 1024
        pass_msg(f"SKILL.md found ({size_kb:.1f} KB)")
    else:
        fail_msg(f"SKILL.md not found at {paths['skill_file']}")

    evals_file = Path(paths["evals_file"])
    if evals_file.exists():
        try:
            raw = json.loads(evals_file.read_text())
            parsed = EvalsFile.model_validate(raw)
            total = sum(len(e.assertions) for e in parsed.evals)
            pass_msg(f"evals.json valid ({len(parsed.evals)} evals, {total} assertions)")
        except Exception as err:
            fail_msg(f"evals.json invalid: {extract_error_message(err)}")
    else:
        fail_msg(f"evals.json not found at {paths['evals_file']}")

    try:
        config = load_config(CliOverrides(config_path=config_path))
        for p in config.providers:
            pass_msg(f"{p.model}: API key valid")
    except Exception as err:
        fail_msg(f"Config: {extract_error_message(err)}")

    console.print()

    if has_error:
        console.print("[red]Validation found issues. Fix them before running evals.[/red]\n")
        raise SystemExit(1)
    else:
        console.print("[green]All checks passed. Ready to run evals.[/green]\n")
